### ISSUE TITLE
Feature: Request for making nodes re-attestable when using  NodeAttestor `aws_iid` plugin

### ISSUE THREAD

**Issue by ranjit-se7en** [spiffe/spire#5190]
#### Feature: Request for making EC2 Instances re-attestable when using `aws_iid` plugin

* **Version**: `SPIRE_VERSION="1.9.4"`
* **Platform**: `Linux 6.1.90-99.173.amzn2023.x86_64 #1 SMP PREEMPT_DYNAMIC Tue May  7 11:11:31 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux` 
* **Subsystem**:  `aws_iid`

We are working on a POC where we would like to run `spire-server` in K8s and `spire-agents` on EC2 instances utilizing the `aws_iid` NodeAttestor plugin.

However, we noticed that whenever the spire-agent crashes/restarted(or the instance itself is restarted), there is no way to re-attest the agents, and the only option available is to evict the nodes from the server in K8s (using kubectl exec). We understand that this is already documented under the plugins [Security Considerations](https://github.com/spiffe/spire/blob/main/doc/plugin_server_nodeattestor_aws_iid.md#security-considerations) section.

```
While many operators choose to configure their systems to block access to the Instance Identity Document, the SPIRE project cannot guarantee this posture. To mitigate the associated risk, the aws_iid node attestor implements Trust On First Use (or TOFU) semantics. For any given node, attestation may occur only once. Subsequent attestation attempts will be rejected.
```

We would like to know if it's possible to make the node re-attestable since it's not feasible to evict the agents every time. We already has a discussion around this in the spire slack channel and looks like it could be made possible.

Thanks!

**Comment by kfox1111** [comment]
Having had some discussions with folks on slack, I think there are some configurations where it may be safe to turn off TOFU with the existing plugin. Maybe we need a new config option for the plugin to make it possible to disable?

**Comment by dfeldman** [comment]
Here is what I have been thinking. I've been bouncing this idea off various people but have nothing implemented yet. 

- New node attestor called "aws_tpmx". Tpmx stands for "TPM with extended metadata." To distinguish from other node attestors that use TPM alone. 
- First, on the agent side, it would determine the instance ID. This would come from IMDS by default, but you could also override it with a configuration option (if IMDS is unavailable). This instance ID is *not trusted* by the server. 
- On the server side, it would query the AWS API to determine the EK of the agent's instance id. 
- The server would generate a nonce and ask the agent to sign the nonce with that EK. 
- The server would verify the nonce. This proves that the instance ID sent by the client earlier is correct.
- The server would set selectors on the node attestor for instance ID, tags, roles, etc. all determined from the AWS API using the instance ID. 

The reason for this idea:
- This allows blocking the IMDS entirely, which is useful for high security environments (and the default for some setups). 
- This exact same approach works on GCP and Azure, which have APIs to get the EK of any instance. 
- The agent is proving that it is the real instance that has the quoted instance ID, AND that it has access to /dev/tpm, which should be only privileged processes running on that instance. 
- The challenge-response with signing a server-generated nonce is reattestable since it cannot be replayed. 

AWS PCR4 register unique behavior:
- In AWS only, instances have a hash of the instance ARN in their TPM PCR4 register. 
- I do not think it is wise to rely on this feature though. First, it is AWS unique behavior. Second, because it's only a hash of the ARN, it cannot be used as an index to look up any details of the instance, only to verify them.

**Comment by kfox1111** [comment]
I think this plan sounds basically like the boxboat tpm plugin, plus some extra selector work added? Maybe that plugin could be extended with the X functionality on top?

**Comment by dfeldman** [comment]
The Boxboat plugin is a good starting point, but I'm fairly sure it doesn't have the step where it uses the AWS API to fetch the EK. (That functionality in AWS is actually very new -- less than 1 month old).

**Comment by kfox1111** [comment]
Right. But that is probably an extension to what the plugin is already doing. like, if you could take the boxboat server plugin, and extend the class with an extra set of calls to the end of attestation, after it has done the tpm part, to add the cloud attestation/selectors, that might simplify the implementation?

**Comment by kfox1111** [comment]
I guess what I'm saying is, maybe a lot of the boxboat plugin could be used to implement the other plugins being discussed rather then start a new one, if done carefully. Maybe a good reason to bring the boxboat plugin in tree.

**Comment by dfeldman** [comment]
Makes sense. The code would be largely common between AWS, GCP, and Azure versions of this node attestor.

**Comment by kfox1111** [comment]
I could even see it being used with other plugins as well. maybe an openstack one, or a vmware one, as they get their own equivalents.

**Comment by azdagron** [comment]
> However, we noticed that whenever the spire-agent crashes/restarted(or the instance itself is restarted), there is no way to re-attest the agents
What KeyManager are your agents using? If the agent is able to persist it's key, then it shouldn't need to re-attest on startup (assuming its SVID has not expired.... which would happen if the agent was down for a long time).

**Comment by kfox1111** [comment]
I do like the idea of having cloud providers that have to periodically re-prove the vm is still where it is via reattestation. Regardless if TOFU can be made to stick for this specific case.

**Comment by ranjit-se7en** [comment]
> > However, we noticed that whenever the spire-agent crashes/restarted(or the instance itself is restarted), there is no way to re-attest the agents
> 
> What KeyManager are your agents using? If the agent is able to persist it's key, then it shouldn't need to re-attest on startup (assuming its SVID has not expired.... which would happen if the agent was down for a long time).

We are using the KeyManager `disk` plugin, and no the agents don't re-attest while SVID is not expired.

**Comment by azdagron** [comment]
I think we'd be fairly hesitant to just add a configuration to the plugin that disabled TOFU. This configuration would be too easy to reach for when facing friction with agent eviction. Disabling TOFU should probably only be done with extreme scrutiny.

Are there situations where SPIRE itself could determine that it was safe to disable TOFU?

> Having had some discussions with folks on slack, I think there are some configurations where it may be safe to turn off TOFU with the existing plugin. Maybe we need a new config option for the plugin to make it possible to disable?

@kfox1111, what configurations do you know about?

**Comment by kfox1111** [comment]
Thinking we might have a new option to have a unix socket based proxy to the aws_iid spire-server can use.

In a kubernetes environment, they are blocking access to the iid by any non hostNetwork: true pod. So, pretty safe. Could probably verify this by just trying to contact it directly from the spire-server.

All the spire-agents are running hostNetwork: true so should be able to access iid. So, nothing special to do there.

We could have a daemonset for a small proxy to make iid available to the servers that are locked down via unix socket. Along with a csi instance making it available to spire-servers. The existing spiffe driver should work without modifications for that. Then the server does not have to run hostNetwork=true and require a lot more permissions in kubernetes.

With aws gaining tpm support, tpms could be switched out for iid on the agent side for better security yet. The spire-server will still need iid access though I think to verify the tpms.

**Comment by kfox1111** [comment]
It helps I think, but doesn't solve one of the root problems. How do you restrict access to the metadata server to just the spire server so you can trust that only it has access to do the handshaking? The tpm route has the advantage of being a /dev/tpm device that you can restrict permissions to.

**Comment by arianvp** [comment]
What I do in production is the following:

I have `IPAddressDeny=169.254.169.254` on `-.slice`  and  `IPAddressAllow=169.254.169.254` on `spire-server.service`

**Comment by kfox1111** [comment]
Very interesting... I've been needing to do something like that for something else....

Do you somehow change the slice of the spire-agent to allow it to skip the deny rule then? -.slice affects things that dont specify a slice?

**Comment by arianvp** [comment]
`-.slice` is the root slice under which all other units live. 

See the docs in https://www.freedesktop.org/software/systemd/man/latest/systemd.resource-control.html

Specifically the docs on `IPAddressAllow/Deny` https://www.freedesktop.org/software/systemd/man/latest/systemd.resource-control.html#IPAddressAllow=ADDRESS%5B/PREFIXLENGTH%5D…



My literal config is:

```
# /etc/systemd/system/-.slice.d/override.conf
# Deny all systemd units access to IMDS
[Slice]
IPAddressDeny=169.254.169.254
```

```
# /etc/systemd/spire-agent.service/override.conf
# Allow spire-agent access to IMDS
[Service]
IPAddressAllow=169.254.169.254
```

**Comment by arianvp** [comment]
(You might also need an IPAddressAllow to access the spire-server iirc. I have to look up my config later when I'm home . I am typing this from memory in my phone)

**Comment by kfox1111** [comment]
the slice there seems to apply to everything, then the agent has an explicit allow too. so it has both deny and allow and the allow takes precidence? or can you scope the agent so the default slice's rules dont take effect? (maybe putting it in a different slice?)

**Comment by arianvp** [comment]
Again see the docs in `man systemd.resource-control`. the rule is implemented as follows:

* Access is granted when the checked IP address matches an entry in the IPAddressAllow= list.
* Otherwise, access is denied when the checked IP address matches an entry in the IPAddressDeny= list.
*  Otherwise, access is granted.

**Comment by kfox1111** [comment]
Ah. ok. thanks for the info. Wont work for what I'm trying to do then. trying to exclude a protocol by all units by default via RestrictAddressFamilies, then add it only to those units that need it. But if its still applying to the specific units, it will still merge in the block, the block has precedent over allow in RestrictAddressFamilies config.

**Comment by kfox1111** [comment]
As for the node attestor use case, I think that may just work. :)

maybe a socat -> unix socket bridge of some sort for spire running in k8s as well.

**Comment by kfox1111** [comment]
I've actually started working on the tpm plugin to support aws. I'm hoping to have it done soon.
